Rename to postal_code/postal code

lcezermf opened the pull request:
Rename to postal_code/postal code

- Rename from zipcode to postal_code in ex/exs files
- Rename from Zip codes/zipcode to postal code in text files

Fix #17

lcezermf committed:
Rename to postal_code/postal code

- Rename from zipcode to postal_code in ex/exs files
- Rename from Zip codes/zipcode to postal code in text files

prodis commented on lib/correios/cep.ex:
```diff
@@ -80,36 +80,36 @@ defmodule Correios.CEP do
       |> parse()
     else
-      {:error, Error.new("zipcode in invalid format")}
+      {:error, Error.new("postal_code in invalid format")}
```
```suggestion
      {:error, Error.new("postal code in invalid format")}
```

prodis commented on lib/correios/cep/address.ex:
```diff
@@ -3,7 +3,7 @@ defmodule Correios.CEP.Address do
   """
 
-  @enforce_keys [:street, :complement, :neighborhood, :city, :state, :zipcode]
+  @enforce_keys [:street, :complement, :neighborhood, :city, :state, :postal_code]
```
As a rule to avoid breaking changes, it would be better to add the new field `postal_code` and keep `zipcode` to be removed in a later release. 

On the other hand, if the communication of the breaking changing is effective (README, CHANGELOG, etc), it should not be a pain point for the consumers of the library just to rename a field in the expected response. 

@lccezinha Are you consuming Correios CEP? What is your opinion from the library's consumer point of view?

lcezermf replied:
Hey, @prodis thanks a lot for your response.

No, I'm not using the library, it just appears to me in the ElixirWeekly 203 and I decide to take a look for issues to contribute.

About the rule to avoid breaking changes, I would keep both fields `zipcode` and `postal_code` for a while, update patch version of the lib and try to add a message to warn about the deprecation of the field `zipcode`.

When you update the minor version I would remove the legacy field `zipcode` to keep only `postal_code`.

prodis replied:
Please do that.

lcezermf replied:
I'll do that :+1:

lcezermf replied:
Hey @prodis sorry about the lack of updates, let me ask you a question. Do you suggest add again the field `zipcode` to the code base and also keep the documentation using it, or the documentation must be already using the new `postal_code` field and the old `zipcode` only will be in the code to make the old version compatible?

prodis replied:
Just add the `zipcode` field back to `Correios.CEP.Address` struct and leave the documentation as it is with `postal_code`. I will update the documentation with the deprecation after merging your pull request.

lcezermf replied:
@prodis my last commit broke some stuff could help me?

prodis replied:
Thanks Dialyzer. 🙌 

https://github.com/prodis/correios-cep-elixir/pull/21#discussion_r436268695

prodis reviewed (commented):
@lccezinha Thank you for the PR, it is look good in general. 👍 

I added a change and left one question. 🙂

lcezermf committed:
Update lib/correios/cep.ex

Co-authored-by: Fernando Hamasaki de Amorim <prodis@gmail.com>

lcezermf committed:
Fix doctests

lcezermf committed:
Enable zipcode to use same value as postal_code

- To keep the old versions working this commit will keep both fields
even if the show the same value
- The scruct new/1 keeps the same interface but the response has one more field: zipcode

prodis commented on lib/correios/cep/address.ex:
```diff
@@ -3,16 +3,16 @@ defmodule Correios.CEP.Address do
-
-  defstruct @enforce_keys
+  @enforce_keys [:street, :complement, :neighborhood, :city, :state]
+  defstruct @enforce_keys ++ [:postal_code, :zipcode]
```
Why not are `postal_code` and `zipcode` in `@enforce_keys`?

lcezermf replied:
Maybe I miss understand; If I keep both fields `postal_code` and `zipcode` it will not make both fields required to be filled when using the struct?

lcezermf replied:
I thought that I was going to need two "constructors", one that allows using zipcode and another one that allows using postal_code.

To be clear, I need to add both fields in the enforce_keys and keep a single `new` function with these keys in the map, but both will fill the value for `postal_code`, is that correct?

prodis replied:
The expected response should have both `postal_code` and `zipcode` fields:
```
iex> Correios.CEP.find_address("54250610")
{:ok,
 %Correios.CEP.Address{
   city: "Jaboatão dos Guararapes",
   complement: "",
   neighborhood: "Cavaleiro",
   postal_code: "54250610",
   state: "PE",
   street: "Rua Fernando Amorim",
   zipcode: "54250610"
 }}
```

That way does not break the current consumers and give them the opportunity to start using `postal_code` instead of the current `zipcode` that would be removed in a future version.

prodis commented on lib/correios/cep/address.ex:
```diff
@@ -23,6 +23,56 @@ defmodule Correios.CEP.Address do
+  end
+
+  @doc """
+  Creates a new `#{inspect(__MODULE__)}` struct with the given values.
```
No need to duplicate the `new/1`, since `zipcode` key will never be present in the input params. 

https://github.com/prodis/correios-cep-elixir/blob/f548949ff57bf774777cd328aabc9ccb612681be/lib/correios/cep/parser.ex#L17

lcezermf replied:
I'll fix it.

prodis commented on lib/correios/cep/address.ex:
```diff
@@ -23,6 +23,56 @@ defmodule Correios.CEP.Address do
+      neighborhood: to_string(neighborhood),
+      city: to_string(city),
+      state: to_string(state),
+      postal_code: to_string(postal_code),
```
It is missing to assign the value of `postal_code` to `zipcode` field.

prodis reviewed (approved):
LGTM

Thanks!
